### DISCUSSION TITLE
Issue with Jellyfin Import when playing media

### DISCUSSION THREAD

**Discussion by FoMoCo53** [arabcoders/watchstate#416] [General] [answered]
Hi,

I have been doing some investigation into an issue I have been having and what I have worked out is if the import process is run whilst I am playing a file in Jellyfin it will cause the status of the file to become invalid.

I have a Plex and Jellyfin running and have Watchstate set to run via cron every hour. Import 0 * and export 5 *. Which if Jellyfin is not playing a file works fine and I have no issues. However if I am playing something in Jellyfin when the import job runs that is when it gets all screwed up.

During investigating the issues, I discovered that Jellyfin had wrong metadata on a bunch of files which caused WS to keep importing the the same files again and again. So I kept running the Metadata only import until I fixed all the errors and it wasnt importing any new data. aka removed all the duplicates etc.

After doing this I was still randomly getting files that didnt import correctly. And after going through the app*.log I suspected the issue and changed the cron job to run at 0100 only. And it seems to be fine

Below are extracts from the app*.log let me know if you need more info
I cant say exactly when I started to watch or finished watching this particular file, and I couldnt work out what the progress number meant. But happy to do further testing if it helps

This is the first entry of this file in the log, I suspect this was whilst I was watching it

    [2024-03-13T18:00:30.824867+11:00] logger.NOTICE: MAPPER: [jellyfin] updated [Ray Donovan (2014) - 02x005] due to play progress change. {"id":13888,"changes":{"metadata":{"jellyfin":{"progress":{"old":"0","new":"650268"}}}},"fields":"metadata"} []

This looks to be next import run and it seems to have marked it as watched, but if I am working this out it correctly has marked it as watch at 2024-03-13 17:51:21, which is before the previous import job ran. It is almost like Jellyfin is marking the file as watched at the time I started watching it?

    [2024-03-13T19:01:14.423510+11:00] logger.NOTICE: MAPPER: [jellyfin] updated [Ray Donovan (2014) - 02x005] due to play progress change. {"id":13888,"changes":{"metadata":{"jellyfin":{"watched":{"old":"0","new":"1"},"progress":{"old":"650268","new":"0"},"played_at":{"old":"None","new":"1710312681"}}}},"fields":"metadata"} []

This is then the next time the import ran and the file is now in an invalid state. It stays like this until I manually fix it. I was originally just marking it as played in Plex, but then I tested marking it as unplayed then played in Jellyfin and this also fixes it, as it updates the watched date to valid one.

    [2024-03-13T20:00:56.476052+11:00] logger.NOTICE: MAPPER: [jellyfin] item [13888: Ray Donovan (2014) - 02x005] is marked as [played] vs local state [unplayed], However due the remote item date [2024-03-13T17:51:21+11:00] being older than the last backend sync date [2024-03-13T19:00:48+11:00]. it was not considered as valid state. [] []

Let me know if you want more information or testing done.

Thanks Heaps for all your work
FoMoCo

**Comment by arabcoders** [comment]
Thanks, This seems to be an issue with jellyfin backend when the media is being played. i thought i marked the event as tainted as such it shouldn't trigger mark as watched, However. I'll have to dig into it, in the meantime, try using the webhooks for updating and see if it fixes the issue for you.

**Comment by arabcoders** [comment]
@FoMoCo53 , i think I've traced the issue, Please try `ghcr.io/arabcoders/watchstate:dev` branch and see if it fixes your problem.

**Reply by arabcoders** [reply]
The fix has been applied to latest branch, feel free to test it.

**Comment by FoMoCo53** [comment]
Hi,

I just finished running a test on the :dev branch and it is doing the same thing.

    [2024-03-16T08:58:36.233394+11:00] logger.NOTICE: MAPPER: [jellyfin] item [13891: Ray Donovan (2014) - 02x008] is marked as [played] vs local state [unplayed], However due the remote item date [2024-03-16T08:10:36+11:00] being older than the last backend sync date [2024-03-16T08:57:14+11:00]. it was not considered as valid state. [] []

    docker exec -ti watchstate console -V
    WatchState 20240315-cb4d493

I took note of the time that I started watching the episode and that is the time that the episode is being marked as played at. Not the time that I actually finished watching the episode. This seems wrong on the behalf of Jellyfin and makes no sense why it is marking the played at time the time you started watching it. Correct me if I am wrong, but I suspect that is what is causing the issue.

Thanks again

FoMoCo

**Reply by arabcoders** [reply]
Hi, i assume the data is stale the fix will work on future data import. Try starting over or force reimport to correct the state in db. 

If it's still happening please re-update with the relevant logs again. 

The issue it seems in play progress code branch accepting metadata related to watch state, which isn't supposed to happen at least with the latest fix. This issue mostly occurs when webhooks are disabled. 

Thanks.

**Reply by FoMoCo53** [reply]
Hi,

The process I took for the testing was as follows

Downed the Container
Updated docker-compose to the :dev branch
Bought up the container and it pulled down the new image
Ran a "console state:import -v" then a "console state:export -v" to ensure that it was up to date. No changes were detected or updates made
Then started playing the episode
Ran a "console state:import -v" whilst it was playing
Waited for the episode to finish playing then ran a "console state:import -v"

    [2024-03-16T08:57:38.897603+11:00] logger.NOTICE: MAPPER: [jellyfin] updated [Ray Donovan (2014) - 02x008] due to play progress change. {"id":13891,"changes":{"metadata":{"jellyfin":{"watched":{"old":"0","new":0},"progress":{"old":"640144","new":"0"},"played_at":{"old":"None","new":"1710537036"}}}},"fields":"metadata"} []

Then ran "console state:import -v" again and this is when it reported an invalid state

    [2024-03-16T08:58:36.233394+11:00] logger.NOTICE: MAPPER: [jellyfin] item [13891: Ray Donovan (2014) - 02x008] is marked as [played] vs local state [unplayed], However due the remote item date [2024-03-16T08:10:36+11:00] being older than the last backend sync date [2024-03-16T08:57:14+11:00]. it was not considered as valid state. [] []

Please confirm the above process is correct or what changes I need to do and will run another test

Thanks
FoMoCo

**Reply by arabcoders** [reply]
Thanks this seems to be correct, did you actually start with new database or you kept the data and just run import over the old data? 

sync date is kept in servers.yaml and db state in the database itself.

**Reply by FoMoCo53** [reply]
That was using the existing data, I just changed the branch on the existing container and ran an import export

Do you want me to try with a new DB? If so what is the best way to do so?

**Reply by arabcoders** [reply]
You try with completely new data remove the ./data folder and start a new 

This is the best way to ensure test data integrity.

**Reply by FoMoCo53** [reply]
I suspected that might be the case.
Ill change to the :latest branch, run through the setup, run a test and let you know how it goes

**Reply by FoMoCo53** [reply]
Running through the setup of plex and got this error

```
Would you like to import play state & metadata from the backend now? [Y|N] [Default: No]
-----------------
P.S: this could take few minutes to execute.

> y
Importing play state & metadata from plex

In ImportCommand.php line 255:

  count(): Argument #1 ($value) must be of type Countable|array, string given


config:add [-s|--select-backend SELECT-BACKEND]
```

ill run a manual import for now

**Reply by arabcoders** [reply]
Indeed it seems i forgot to update this code when we changed how we handle some arguments. No worries you can safely ignore this error and do the manual import.

**Comment by FoMoCo53** [comment]
I just ran another test after doing a fresh setup and it is still doing the same thing

Started playing at 11:07

A few minutes into playing

    console state:import -v
    [2024-03-16T11:16:03.792997+11:00] logger.NOTICE: Using WatchState Version - '20240315-2dd9c08'. [] []
    [2024-03-16T11:16:30.830957+11:00] logger.NOTICE: MAPPER: [jellyfin] updated [Ray Donovan (2014) - 02x009] due to play progress change. {"id":13903,"changes":{"metadata":{"jellyfin":{"watched":{"old":"0","new":0},"progress":{"old":"0","new":"490345"}}}},"fields":"metadata"} []

Finished playing

    console state:import -v
    [2024-03-16T11:52:49.782488+11:00] logger.NOTICE: MAPPER: [jellyfin] updated [Ray Donovan (2014) - 02x009] due to play progress change. {"id":13903,"changes":{"metadata":{"jellyfin":{"watched":{"old":"0","new":0},"progress":{"old":"490345","new":"0"},"played_at":{"old":"None","new":"1710547631"}}}},"fields":"metadata"} []

Re Ran

    console state:import -v
    [2024-03-16T11:55:04.522467+11:00] logger.NOTICE: MAPPER: [jellyfin] item [13903: Ray Donovan (2014) - 02x009] is marked as [played] vs local state [unplayed], However due the remote item date [2024-03-16T11:07:11+11:00] being older than the last backend sync date [2024-03-16T11:52:24+11:00]. it was not considered as valid state. [] []

FoMoCo

**Reply by arabcoders** [reply]
Thank you for confirming the issue. I believe i might have noticed the issue. Hopefully will have a fix tomorrow. In the meantime could you run system:report for me and post the results.

**Comment by FoMoCo53** [comment]
Thanks again for looking into it here is the report minus the plex keys

```
[ Basic Report ]

WatchState Version: 20240315-2dd9c08
PHP Version: 8.2.17
Timezone: Australia/Sydney
Running in Container? Yes
Data Path: /config
Temp Path: /config
Database Migrated?: Yes
Has .env file? No
Is Tasks Runner working? Yes
Report Generated At: 2024-03-16T01:04:26+00:00

[ Backends ]

[ Plex ==> plex ]

Is backend URL HTTPS? Yes
Has Unique Identifier? Yes
Has User? Yes
Export Enabled? Yes
Time since last export? 2024-03-16T00:04:22+00:00
Play state import enabled? Yes
Metadata only import enabled? No
Time since last import? 2024-03-16T00:54:39+00:00
Has webhook token? Yes
Is webhook match user id enabled? Yes
Is webhook match backend unique id enabled? Yes
Has custom options? Yes
{"plex_user_uuid":"xxxxxxx","ADMIN_TOKEN":"xxxxxxx"}
[ Jellyfin ==> jellyfin ]

Is backend URL HTTPS? Yes
Has Unique Identifier? Yes
Has User? Yes
Export Enabled? Yes
Time since last export? 2024-03-16T00:04:22+00:00
Play state import enabled? Yes
Metadata only import enabled? No
Time since last import? 2024-03-16T00:54:39+00:00
Has webhook token? Yes
Is webhook match user id enabled? No
Is webhook match backend unique id enabled? Yes
Has custom options? No
{}
[ Tasks ]

[ Import ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? 0 1 * * *
When is the next scheduled run? 2024-03-16T05:00:00+00:00

[ Export ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? 5 1 * * *
When is the next scheduled run? 2024-03-16T05:05:00+00:00

[ Push ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? */10 * * * *
When is the next scheduled run? 2024-03-16T01:10:00+00:00

[ Progress ]

Is Task enabled? No

[ Backup ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? 0 6 */3 * *
When is the next scheduled run? 2024-03-18T19:00:00+00:00

[ Prune ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? 0 */12 * * *
When is the next scheduled run? 2024-03-16T13:00:00+00:00

[ Indexes ]

Is Task enabled? Yes
Which flags are used to run the task? -v
When the task scheduled to run at? 0 3 * * 3
When is the next scheduled run? 2024-03-19T16:00:00+00:00

[ Requests ]

Is Task enabled? Yes
Which flags are used to run the task? -v --no-stats
When the task scheduled to run at? */2 * * * *
When is the next scheduled run? 2024-03-16T01:06:00+00:00

[ Logs ]

[ /config/logs/app.20240316.log ]

[2024-03-16T11:52:56.000726+11:00] logger.NOTICE: SYSTEM: Found [14] updated objects. {"memory":{"now":"233.44M","peak":"311.64M"}} []
[2024-03-16T11:54:37.623872+11:00] logger.NOTICE: Using WatchState Version - '20240315-2dd9c08'. [] []
[2024-03-16T11:54:37.624983+11:00] logger.NOTICE: SYSTEM: Preloading MemoryMapper data. {"memory":{"now":"4.58M","peak":"4.60M"}} []
[2024-03-16T11:54:38.824991+11:00] logger.NOTICE: SYSTEM: Preloading MemoryMapper data is complete. {"memory":{"now":"231.81M","peak":"232.30M"}} []
[2024-03-16T11:54:38.827068+11:00] logger.NOTICE: SYSTEM: Importing [plex] METADATA & PLAY STATE changes. {"since":"2024-03-16 11:52:24 AEDT"} []
[2024-03-16T11:54:39.591826+11:00] logger.NOTICE: SYSTEM: Importing [jellyfin] METADATA & PLAY STATE changes. {"since":"2024-03-16 11:52:24 AEDT"} []
[2024-03-16T11:54:39.657161+11:00] logger.NOTICE: SYSTEM: Waiting on [47] requests. {"time":{"start":"2024-03-16T11:54:39+11:00"},"memory":{"now":"234.03M","peak":"234.03M"}} []
[2024-03-16T11:55:04.522467+11:00] logger.NOTICE: MAPPER: [jellyfin] item [13903: Ray Donovan (2014) - 02x009] is marked as [played] vs local state [unplayed], However due the remote item date [2024-03-16T11:07:11+11:00] being older than the last backend sync date [2024-03-16T11:52:24+11:00]. it was not considered as valid state. [] []
[2024-03-16T11:55:10.672156+11:00] logger.NOTICE: SYSTEM: Finished waiting on [47] requests. {"time":{"start":"2024-03-16T11:54:39+11:00","end":"2024-03-16T11:55:10+11:00","duration":31},"memory":{"now":"233.44M","peak":"311.45M"},"responses":{"size":"61.40M"}} []
[2024-03-16T11:55:10.672660+11:00] logger.NOTICE: MAPPER: No changes detected. [] []

[ /config/logs/access.20240316.log ]

access log file is empty or does not exists.

[ /config/logs/task.20240316.log ]

task log file is empty or does not exists.

[ /config/logs/app.20240315.log ]

app log file is empty or does not exists.

[ /config/logs/access.20240315.log ]

access log file is empty or does not exists.

[ /config/logs/task.20240315.log ]

task log file is empty or does not exists.

[ Notice ]

Beware, while we try to make sure no sensitive information is leaked, it's possible
that some private information might be leaked via the logs section.
Please review the report before posting it.
```

**Reply by arabcoders** [reply]
I attempted to trigger the bug many times, it's just not triggering

i tried
* while playing
* while stopped
* while at the end

![image](https://github.com/arabcoders/watchstate/assets/1621552/570f2a4e-cda8-4360-89e4-4bde917e9e5b)

it's just not happening are you perhaps on jellyfin 10.9?

**Reply by arabcoders** [reply]
Nevermind, i was able to trigger the bug

**Comment by arabcoders** [comment]
Ok, After deep testing it seems, jellyfin does not update the `UserData.LastPlayedDate` when the item marked as played. And due to the way we handle item progress update via `state:import` it was causing conflict. Sadly there is no easy way to fix it with the current approach. Thus, i am forced to disable item progress update via `state:import` for jellyfin clients and rely solely on webhooks until such time i can re-write the logic handling for jellyfin or they fix the bug and update the `LastPlayedDate` correctly.

Thank you for your report and logs. I've pushed new build to dev which should be up for latest it will fix the issue temporary as described above.

**Comment by FoMoCo53** [comment]
Yeah I suspected something like that might be the case. Thanks for looking into it. Let me know if you want me to do more testing in the future.

**Reply by arabcoders** [reply]
Please do try the latest version and see if you still get errors.

**Reply by FoMoCo53** [reply]
I just ran a test on version `20240316-222e200` from the :latest branch

In the logs there is no longer mentions of the progress state of the file whilst it is playing. And when I finished playing and it has been marked as watched by JF it is still reporting the same invalid state on the next import run, due to the `LastPlayedDate` being set to when the file was started playing and not finished.

I dont know the full logic behind why the testing for a valid state is currently being done the way it is or what values are available to test against. But would testing the new `LastPlayedDate` against the last played date for that file in the DB work? Yes there might be a conflicts if Plex/Emby are setting the `LastPlayedDate` to the time the file was finished being watched. This would only happen if you were watching the same file on two servers at once (unlikely) and ultimately would result in the file being marked as watched in all systems, which is the desired result yes? Just thinking out aloud so might be way off base

**Reply by arabcoders** [reply]
> I just ran a test on version `20240316-222e200` from the :latest branch
> 
> 
> 
> In the logs there is no longer mentions of the progress state of the file whilst it is playing. And when I finished playing and it has been marked as watched by JF it is still reporting the same invalid state on the next import run, due to the `LastPlayedDate` being set to when the file was started playing and not finished.
> 
> 
> 
> I dont know the full logic behind why the testing for a valid state is currently being done the way it is or what values are available to test against. But would testing the new `LastPlayedDate` against the last played date for that file in the DB work? Yes there might be a conflicts if Plex/Emby are setting the `LastPlayedDate` to the time the file was finished being watched. This would only happen if you were watching the same file on two servers at once (unlikely) and ultimately would result in the file being marked as watched in all systems, which is the desired result yes? Just thinking out aloud so might be way off base
> 
> 

I am on my phone so i cNt explain why the code doing what is doing, once i am back in pc i will write an explanation. 

In the meantime this result you got was based on only test data yeah? not newly generated as explained in the previous post?

**Reply by FoMoCo53** [reply]
That's cool, only explain if you feel like it, I figured there was a more complex reason why the validation test is the way it is.

I am only using a single container. So it is using the same data from when I 'rebuilt' it yesterday. I ran the same process as did previously. Updated container to the new image, import, export to ensure everything was up to date, started playing video, ran import, finished watching video, ran import.

This is sounding like it is going to be a big job to fix, so I am happy for it to join the the queue of bugs/features that you are working on. I understand why it is happening now, and am happy to just run the sync overnight when nothing is playing. I dont jump back and forth between the servers that often.

Thanks again for looking into it and always happy to help with testing etc in the future
FoMoCo

**Reply by arabcoders** [reply]
The logic if you can follow it can be found in [Import Mapper](https://github.com/arabcoders/watchstate/tree/master/src/Libs/Mappers/Import), The way the code structured is to support allowing mark as unplayed, as some backends, mainly plex does not give us many options to support it. The logic which is causing problem for your particular case is https://github.com/arabcoders/watchstate/blob/222e200f7b2856e91b9bb5d4274ab4e851ede861/src/Libs/Mappers/Import/MemoryMapper.php#L254-L359 Which my recent fix i assumed it's fixed at lesat it did for my testing.

I would like you to try again using fresh data, i.e. remove the data directory and try again. and see if you can trigger the error. As i have tested many conditions non triggered the bug in the recent fixes.

Sorry to have wasted your time, but such is the nature of working on bridging incompatible APIs.

**Reply by FoMoCo53** [reply]
Hi, I just got through running the test after deleting the data directory and it is still doing the same thing.
Using WatchState Version - '20240316-222e200'

Oh and to answer a question you asked before that I just realised I didnt answer I am running Jellyfin v10.8.13 and Plex v1.40.1.8227

And you arent wasting my time, I am happy to help and run tests. Thank you for spending the time on the issue and looking into it.

**Reply by arabcoders** [reply]
Well.. this is embarrassing, i probably shouldn't do code updates when fasting lol. I think i truly fixed it now. I had the logic inversed.

I have pushed a fix, please do pull the dev branch and test when you can. try different episode

Thank you for your time.

**Reply by FoMoCo53** [reply]
Hi,

I just updated to the :dev branch and it is back to reporting the progress and still failing

    Using WatchState Version - '20240318-bd1a7ab'.

Running the import whilst playing

    [2024-03-19T08:20:32.800500+11:00] logger.NOTICE: MAPPER: [jellyfin] updated [Ray Donovan (2014) - 02x011] due to play progress change. {"id":13820,"changes":{"metadata":{"jellyfin":{"progress":{"old":"0","new":"266781"}}}},"fields":"metadata"} []

Running the import after playing

    [2024-03-19T08:21:37.748498+11:00] logger.NOTICE: MAPPER: [jellyfin] updated [Ray Donovan (2014) - 02x011] due to play progress change. {"id":13820,"changes":{"metadata":{"jellyfin":{"watched":{"old":"0","new":"1"},"progress":{"old":"266781","new":"0"},"played_at":{"old":"None","new":"1710796795"}}}},"fields":"metadata"} []

Running the import again

    [2024-03-19T08:22:15.165189+11:00] logger.NOTICE: MAPPER: [jellyfin] item [13820: Ray Donovan (2014) - 02x011] is marked as [played] vs local state [unplayed], However due the remote item date [2024-03-19T08:19:55+11:00] being older than the last backend sync date [2024-03-19T08:21:13+11:00]. it was not considered as valid state. [] []

**Reply by arabcoders** [reply]
Ok, Thank you for your update. It seems i really have to completely disable progress handling for jellyfin at all levels.

I'll get back to you hopefully tomorrow.

**Reply by arabcoders** [reply, marked as answer]
@FoMoCo53, Ok i had time to sit down and trace the issue, it's as you guessed, Jellyfin is not updating the date on object, which in turn confuses the importer, sadly there is nothing to fix as it's worked as intended. 

However, I've implemented workaround for this issue which sadly only works via webhooks. Enable webhooks for jellyfin, and it should work as expected temporarily until someone raises an issue with jellyfin devs. 

Please take a moment to test this and report back.
